[Doc] advanced example for simulating proppant slot test (#1573)
GEOSX/GEOSX: src/docs/sphinx/advancedExamples/validationStudies/proppantSlotTest/Example.rst

.. _ExampleProppantTest:

####################################################

> Simulate Proppant Slot Test

Reviewer: ```suggestion
Proppant Slot Test
```

####################################################

**Context**

> In this example, a simulation is built up to model proppant slot test. In this way, the implemented proppant model is validated by comparing numerical results with the corresponding experimental data. Furthermore, this calibrated proppant model can allow field engineers to customize stimulation design and optimize field operations in multiple engineering aspects `(Huang et al., 2021) <https://onepetro.org/ARMAUSRMS/proceedings/ARMA21/All-ARMA21/ARMA-2021-1248/467980>`__.

Reviewer: ```suggestion
In this example, a simulation is built up to model a proppant slot test. In this way, the implemented proppant model is validated by comparing numerical results with the corresponding experimental data. Furthermore, this calibrated proppant model can allow field engineers to customize stimulation design and optimize field operations in multiple engineering aspects `(Huang et al., 2021) <https://onepetro.org/ARMAUSRMS/proceedings/ARMA21/All-ARMA21/ARMA-2021-1248/467980>`__.
```

**Input file**

> This example uses no external input files and everything required is

Reviewer: ```suggestion
This example uses no external input files and everything is
```

contained within a single xml file that is located at:

.. code-block:: console

  src/coreComponents/physicsSolvers/multiphysics/benchmarks/Proppant-transport-validation-case.xml

------------------------------------------------------------------
Description of the case
------------------------------------------------------------------

> `Chun et al. (2020) <https://www.sciencedirect.com/science/article/pii/S0920410519309441>`__ conducted slot tests on proppant transport with slickwater. As shown below, a 4 ft X 1 ft slot with 0.3 in (0.00762 m) gap width was constructed. Three fluid inlets with 0.5 in inner diameter were placed at the right side of the slot, which were three inches away from each other. One outlet was place on the top side to allow pressure relief. The other one was located on the left side acting as a fluid sink. In their test, to resemble a slickwater fracturing treatment, the proppant concentration was kept at 1.5 ppg and the viscosity of carrying fluid was approximately 1 cp. The slurry was mixed well and then injected into the flow channel at a constant injection rate (6 gpm). A simulation case with the same settings is built up to mimic this slot test. A vertical and impermeable fracture surface is assumed in this case, which eliminates the effect of fracture plane inclination and fluid leakoff. A static fracture with uniform aperture (0.3 in) is defined and fracture propagation is not involved. 30/50 mesh proppant is injected via the three inlets and is flowed through the slot for 30 seconds.

Reviewer: Maybe in this description we could use the GEOSX unit system everywhere (meters, Pa.s, etc)?

.. _problemSketchFig:
.. figure:: config.png
   :align: center
   :width: 500
   :figclass: align-center

   Configuration of the slot for proppant transport experiment (after `Chun et al., 2020 <https://www.sciencedirect.com/science/article/pii/S0920410519309441>`__)

To simulate proppant transport phenomenon, a proppant solver based on the assumption of multi-component single phase flow is used in this example. Proppant concentration and distribution within the slot are numerically calculated by solving the equations of proppant transport in hydraulic fractures. These numerical predictions are then compared with the corresponding testing results `(Chun et al., 2020)  <https://www.sciencedirect.com/science/article/pii/S0920410519309441>`__ to check the accuracy of modelling results.

In this example, we focus our attention on the ``Solver`` tags, the ``Constitutive`` tags, and the ``FieldSpecifications`` tags.

------------------------------
Mesh
------------------------------

> Following figure shows the generated mesh that is used for solving this problem.

Reviewer: ```suggestion
The following figure shows the mesh used for solving this problem.
```

.. _problemSketchFig:
.. figure:: mesh.png
   :align: center
   :width: 500
   :figclass: align-center

   Generated mesh

> We use the internal mesh generator ``InternalWellbore`` to create a computational domain. 

Reviewer: ```suggestion
We use the internal mesh generator ``InternalMesh`` to create a computational domain. 
```

This mesh contains 2 x 97 x 24 eight-node brick elements in the x, y and z directions, respectively. 
Here, a structured three-dimensional mesh is generated with ``C3D8`` as the elementTypes (eight-node hexahedral elements). 
This mesh is defined as a cell block with the name ``cb1``.

.. literalinclude:: ../../../../../../src/coreComponents/physicsSolvers/multiphysics/benchmarks/Proppant-transport-validation-case.xml
  :language: xml
  :start-after: <!-- SPHINX_PROPPANT_MESH -->
  :end-before: <!-- SPHINX_PROPPANT_MESH_END -->

---------------------------

> Proppant Transport solver

Reviewer: ```suggestion
Proppant transport solver
```
to be consistent with the capitalization used in the other advanced examples

Note that end-users should give each single-physics solver a meaningful and distinct name, as GEOSX will recognize these single-physics solvers based on their customized names and create user-expected coupling.

As demonstrated in this example, to setup a coupled proppant transport solver, we need to define three different solvers in the XML file:

- the proppant transport solver for the fracture region, a solver of type ``ProppantTransport`` called here ``ProppantTransport`` (more information here: :ref:`ProppantTransport`),

.. literalinclude:: ../../../../../../src/coreComponents/physicsSolvers/multiphysics/benchmarks/Proppant-transport-validation-case.xml
  :language: xml
  :start-after: <!-- SPHINX_PROPPANT_PROPPANTSOLVER -->
  :end-before: <!-- SPHINX_PROPPANT_PROPPANTSOLVER_END -->

- the single-phase flow solver, a solver of type ``SinglePhaseProppantFVM`` called here ``SinglePhaseFVM``,

.. literalinclude:: ../../../../../../src/coreComponents/physicsSolvers/multiphysics/benchmarks/Proppant-transport-validation-case.xml
  :language: xml
  :start-after: <!-- SPHINX_PROPPANT_FLOWSOLVER -->
  :end-before: <!-- SPHINX_PROPPANT_FLOWSOLVER_END -->

> - the coupling solver (``FlowProppantTransport``) that will bind the two single-physics solvers above, which is named as ``FlowProppantTransport``

Reviewer: ```suggestion
- the coupling solver (``FlowProppantTransport``) that binds the two single-physics solvers above, which is named as ``FlowProppantTransport``
```

.. literalinclude:: ../../../../../../src/coreComponents/physicsSolvers/multiphysics/benchmarks/Proppant-transport-validation-case.xml
  :language: xml
  :start-after: <!-- SPHINX_PROPPANT_COUPLEDSOLVER -->
  :end-before: <!-- SPHINX_PROPPANT_COUPLEDSOLVER_END -->

In this example, let us focus on the coupling solver.

> This solver (``FlowProppantTransport``) describe the coupling process between proppant and flow transport within the ``Fracture`` region. In this way, the two single-physics solvers (``ProppantTransport`` and ``SinglePhaseFVM``) will be simultaneously called and executed for solving this proppant transport problem.

Reviewer: ```suggestion
This solver (``FlowProppantTransport``) describes the coupling process between proppant and flow transport within the ``Fracture`` region. In this way, the two single-physics solvers (``ProppantTransport`` and ``SinglePhaseFVM``) are sequentially called to solve the sub-problems (proppant transport and pressure problem, respectively) involved in this test case.
```

------------------------------
Constitutive laws
------------------------------

> For this slot test, 30/50 mesh proppant was injected via the three inlets and flowing through the slot for 30 seconds. 

Reviewer: ```suggestion
For this slot test, 30/50 mesh proppant is injected via the three inlets and flowing through the slot for 30 seconds. 
```
and I suggest explaining what 30/50 mesh proppant is (I admit I do not know what it means)

> The viscosity of carrying fluid was 1 cp to resemble slickwater fracturing. In this example, the solid and fluid materials are named as ``sand`` and ``water`` respectively. 

Reviewer: ```suggestion
The viscosity of carrying fluid is 0.001 Pa.s to resemble slickwater fracturing. In this example, the solid and fluid materials are named as ``sand`` and ``water`` respectively. 
```

Proppant characterization and fluid rheology are specified in the ``Constitutive`` section: 

.. literalinclude:: ../../../../../../src/coreComponents/physicsSolvers/multiphysics/benchmarks/Proppant-transport-validation-case.xml
    :language: xml
    :start-after: <!-- SPHINX_PROPPANT_MATERIAL -->
    :end-before: <!-- SPHINX_PROPPANT_MATERIAL_END -->

The constitutive parameters such as proppant density and proppant diameter are specified in the International System of Units.

-----------------------------------------------------------
Initial and boundary conditions
-----------------------------------------------------------

The next step is to specify fields, including:

  - The initial value (fracture aperture, fluid pressure and proppant concentration within the fracture have to be initialized)
  - The boundary conditions (fluid pressure and proppant concentration at fluid inlets and outlets)
 
These boundary conditions are set up through the ``FieldSpecifications`` section.

> At a constant injection rate, the slurry was equally flowing into the open channel through three inlets.

Reviewer: ```suggestion
At a constant injection rate, the slurry is equally flowing into the open channel through three inlets.
```

.. literalinclude:: ../../../../../../src/coreComponents/physicsSolvers/multiphysics/benchmarks/Proppant-transport-validation-case.xml
    :language: xml
    :start-after: <!-- SPHINX_PROPANT_BC -->
    :end-before: <!-- SPHINX_PROPPANT_BC_END -->

Note: For static (non-propagating) fracture problems, the fields ``ruptureState`` and ``elementAperture`` should be provided in the initial conditions.

The parameters used in the simulation are summarized in the following table.

+------------------+----------------------------------+------------------+---------------+
| Symbol           | Parameter                        | Unit             | Value         |
+==================+==================================+==================+===============+

> | :math:`d_p`      | Proppant Diameter                | [mm]             | 0.45          |

Reviewer: ```suggestion
| :math:`d_p`      | Proppant Diameter                | [m]             | 0.00045          |
```

+------------------+----------------------------------+------------------+---------------+
| :math:`f`        | Darcy Friction Coefficient       | [-]              | 0.04          |
+------------------+----------------------------------+------------------+---------------+
| :math:`N_{sh}`   | Critical Shields Number          | [-]              | 0.0           |
+------------------+----------------------------------+------------------+---------------+
| :math:`c_s`      | Max Fraction of Proppant         | [-]              | 0.62          |
+------------------+----------------------------------+------------------+---------------+
| :math:`\rho_f`   | Fluid Density                    | [kg/m^3]         | 1000          |
+------------------+----------------------------------+------------------+---------------+
| :math:`\mu_f`    | Fluid Viscosity                  | [Pa*s]           | 0.001         |
+------------------+----------------------------------+------------------+---------------+
| :math:`\rho_p`   | Proppant Density                 | [kg/m^3]         | 2550          |
+------------------+----------------------------------+------------------+---------------+

> | :math:`\lamda_s` | Hindered Settling Coefficient    | [-]              | 4.5           |

Reviewer: ```suggestion
| :math:`\lambda_s` | Hindered Settling Coefficient    | [-]              | 4.5           |
```

       ax[1,0].set_xlabel('Time (second)', weight="bold")
       ax[1,0].set_xlim(0, 40)
       ax[1,0].set_ylim(0.0, 0.6)
       ax[1,0].legend(frameon=False,loc='upper right',fontsize=12)
       ax[1,0].text(.01,.92,'(c)', horizontalalignment='left', transform=ax[1,0].transAxes, size = 15, weight = 'bold')

       ax[1,1].plot(eTime, eArea+esArea, 'ro', label='Experiment-30/50')
       ax[1,1].plot(dTime, (dArea+sArea)/totalCellNum, lw=5, alpha=0.5, color= 'lime', label='GEOSX-30/50 mesh', linestyle='-')
       ax[1,1].set_ylabel('Normalized Total Propped Area', multialignment='center', weight="bold")
       ax[1,1].set_xlabel('Time (second)', weight="bold")
       ax[1,1].set_xlim(0, 40)
       ax[1,1].set_ylim(0.0, 1.0)
       ax[1,1].legend(frameon=False,loc='upper right',fontsize=12)
       ax[1,1].text(.01,.92,'(d)', horizontalalignment='left', transform=ax[1,1].transAxes, size = 15, weight = 'bold')

       plt.show()

   if __name__ == "__main__":
                main()

------------------------------------------------------------------
To go further
------------------------------------------------------------------

**Feedback on this example**

> This concludes the example on simulating proppant slot test.

Reviewer: ```suggestion
This concludes the example on simulating a proppant slot test.
```

| :math:`\lambda_s`| Hindered Settling Coefficient    | [-]              | 4.5           |
+------------------+----------------------------------+------------------+---------------+
| :math:`c_p`      | Proppant Concentration in Slurry | [m^3/m^3]        | 0.07          |
+------------------+----------------------------------+------------------+---------------+
| :math:`L`        | Fracture Length                  | [m]              | 1.219         |
+------------------+----------------------------------+------------------+---------------+
| :math:`H`        | Fracture Height                  | [m]              | 0.3048        |
+------------------+----------------------------------+------------------+---------------+
| :math:`a`        | Fracture Aperture                | [m]              | 0.00762       |
+------------------+----------------------------------+------------------+---------------+
| :math:`Q`        | Injection Rate                   | [m^3/s]          | 0.0003785     |
+------------------+----------------------------------+------------------+---------------+

---------------------------------
Inspecting results
---------------------------------

> The following figure shows the modelling prediction of proppant distribution at t=10s and t=30s, which are compared with the experiments in `(Chun et al., 2020)  <https://www.sciencedirect.com/science/article/pii/S0920410519309441>`__. Due to proppant settling in low viscosity fluid, a heterogeneous proppant distribution is obtained, which evolves with injection time. Three different zones (immobile proppant bed, suspended proppant and clean fluid) are visually identified for both the presented experiment and simulation. 

Reviewer 2: ```suggestion
The following figure shows the modelling prediction of proppant distribution at 10 s and 30 s, which are compared with the experiments in `(Chun et al., 2020)  <https://www.sciencedirect.com/science/article/pii/S0920410519309441>`__. Due to proppant settling in low viscosity fluid, a heterogeneous proppant distribution is obtained, which evolves with injection time. Three different zones (immobile proppant bed, suspended proppant and clean fluid) are visually identified for both the presented experiment and simulation. 
```

.. plot::

   import sys
   import matplotlib
   import numpy as np
   import matplotlib.pyplot as plt

   def main():
       eTime, eLength, eArea = np.loadtxt("experiment-data.txt", skiprows=0, unpack=True)
       esTime, esArea = np.loadtxt("experiment-data2.txt", skiprows=0, unpack=True)

       dTime, dLength, dArea = np.loadtxt("model-results.txt", skiprows=1, unpack=True)
       sTime, sArea = np.loadtxt("model-results2.txt", skiprows=1, unpack=True)
       
       # offSize is the size of the bounday cell which should be excluded from the model domain
       offSize = 0.0127

       # slot length 4 feet = 1.2192 m
       totalLength = 1.2192

       #total cell number of the domain (excluding the bounday cells)  
       totalCellNum = 96 * 24
       font = {'size'   : 12}
       matplotlib.rc('font', **font)
 
       fig, ax = plt.subplots(2,2,figsize=(12, 8))

       ax[0,0].plot(eTime, eLength, 'ro', label='Experiment-30/50')
       ax[0,0].plot(dTime, (dLength-offSize)/totalLength, lw=5, alpha=0.5, color= 'lime', label='GEOSX-30/50 mesh', linestyle='-')
       ax[0,0].set_ylabel('Normalized Bank Length', multialignment='center', weight="bold")

>        ax[0,0].set_xlabel('Time (second)', weight="bold")

Reviewer 2: ```suggestion
       ax[0,0].set_xlabel('Time (s)', weight="bold")
```

       ax[0,0].set_xlim(0, 40)
       ax[0,0].set_ylim(0.0, 1.2)
       ax[0,0].legend(frameon=False,loc='lower right',fontsize=12)
       ax[0,0].text(.01,.92,'(a)', horizontalalignment='left', transform=ax[0,0].transAxes, size = 15, weight = 'bold')

       ax[0,1].plot(eTime, eArea, 'ro', label='Experiment-30/50')
       ax[0,1].plot(dTime, dArea/totalCellNum, lw=5, alpha=0.5, color= 'lime', label='GEOSX-30/50 mesh', linestyle='-')
       ax[0,1].set_ylabel('Normalized Bank Area', multialignment='center', weight="bold")

>        ax[0,1].set_xlabel('Time (second)', weight="bold")

Reviewer 2: ```suggestion
       ax[0,1].set_xlabel('Time (s)', weight="bold")
```

       ax[0,1].set_xlim(0, 40)
       ax[0,1].set_ylim(0.0, 0.5)
       ax[0,1].legend(frameon=False,loc='lower right',fontsize=12)
       ax[0,1].text(.01,.92,'(b)', horizontalalignment='left', transform=ax[0,1].transAxes, size = 15, weight = 'bold')

       ax[1,0].plot(esTime, esArea, 'ro', label='Experiment-30/50')
       ax[1,0].plot(sTime, sArea/totalCellNum, lw=5, alpha=0.5, color= 'lime', label='GEOSX-30/50 mesh', linestyle='-')
       ax[1,0].set_ylabel('Normalized Suspended Proppant Area', multialignment='center', weight="bold")

>        ax[1,0].set_xlabel('Time (second)', weight="bold")

Reviewer 2: ```suggestion
       ax[1,0].set_xlabel('Time (s)', weight="bold")
```

       ax[1,0].set_xlim(0, 40)
       ax[1,0].set_ylim(0.0, 0.6)
       ax[1,0].legend(frameon=False,loc='upper right',fontsize=12)
       ax[1,0].text(.01,.92,'(c)', horizontalalignment='left', transform=ax[1,0].transAxes, size = 15, weight = 'bold')

       ax[1,1].plot(eTime, eArea+esArea, 'ro', label='Experiment-30/50')
       ax[1,1].plot(dTime, (dArea+sArea)/totalCellNum, lw=5, alpha=0.5, color= 'lime', label='GEOSX-30/50 mesh', linestyle='-')
       ax[1,1].set_ylabel('Normalized Total Propped Area', multialignment='center', weight="bold")

>        ax[1,1].set_xlabel('Time (second)', weight="bold")

Reviewer 2: ```suggestion
       ax[1,1].set_xlabel('Time (s)', weight="bold")
```
